File: 3394-minimum-array-end/minimum-array-end.py
class Solution:
    def minEnd(self, n: int, x: int) -> int:
        res = x
        i = 1
        j = 1
        while j <= n - 1:
            if i & x == 0:
                if j & (n - 1):
                    res = res | i
                j  = j << 1
            i = i << 1
        return res 

        # kind off brute force
        # result = x
        # for _ in  range(n - 1):
        #     result += 1
        #     result = result | x
        # return result

File: 3394-minimum-array-end/minimum-array-end.cpp
class Solution {
public:
    long long minEnd(int n, int x) {
        long long res = x, i_x = 1, i_n = 1;
        while (i_n <= n - 1){
            cout << res << " " << i_x << " " << i_n << " " << "\n"; 
            if ((i_x & x) == 0){
                if ((i_n & (n - 1)) > 0){
                    res = (res | i_x);
                }
                i_n = i_n << 1;
            }
            i_x = i_x << 1;   
        }
        return res;

        // for loop
        // long long res = x;
        // for (int i = 0; i < n - 1; i++)
        //     res = (res + 1) | x;
        // return res;
    }
};